### PR TITLE
Fixed incorrect code in metrics 

### PR DESCRIPTION
**Pull request by peastman**
The metrics code had a lot of hardcoded assumptions that are no longer valid about the shapes and types of its inputs.  I've tried to make the handling cleaner and more consistent.  Here are the assumptions it now makes.

1. `y_pred` should generally have 2 dimensions for regression metrics and 3 dimension for classification metrics.  But if there's one less, it interprets that as meaning there's only a single task and fills in the missing task dimension.

2. For classification metrics, `y_true` is generally allowed to contain either class labels or one-hot encoding.  It will automatically convert labels to one-hot if necessary.

3. It's the job of each individual metric function to do any necessary format conversion.  `compute_metric()` should not contain special case handling for particular metrics.  (Though I left in the special handling for `compute_energy_metric`, since I have no idea what that's about!  Someone really needs to clean that up too.)

### CODE REVIEW

File: deepchem/metrics/__init__.py [hunk 1]
```diff
@@ -39,22 +38,29 @@ def from_one_hot(y, axis=1):
...
+def _ensure_class_labels(y):
+  """If necessary, convert one-hot encoding to class labels."""
+  if len(y.shape) == 2:
+    return from_one_hot(y)
+  return y
+
+
+def roc_auc_score(y, y_pred):
+  """Area under the receiver operating characteristic curve."""
+  return sklearn.metrics.roc_auc_score(_ensure_one_hot(y), y_pred)
```
Comment (rbharath) [inline deepchem/metrics/__init__.py]: I like this style!

**Comment by peastman** [comment]
It looks to me like the tests that are failing were actually completely broken before.  The arrays being passed into `compute_metric()` were being completely misinterpreted.  It was just a coincidence they were being misinterpreted in a way that caused the test to pass.

In this test, the model has multiple outputs.  That means that `y_pred` is a list of arrays, one for each output.  `y_true`, on the other hand, is a single array that has been mangled in strange ways by `compute_model_performance()`.  How should `compute_metric()` behave for models with multiple outputs?  Is this even a case it should support?

**Comment by peastman** [comment]
I also don't understand why the tests are using Databag, or what the expected behavior is supposed to be.  In `test_compute_model_performance_singletask_classifier()`, for example, the model is declared to have a single label input of shape `(None, 2)`.  But the value that actually gets passed in has shape `(1, 20, 2)`.  How is that extra dimension supposed to be interpreted?

**Comment by peastman** [comment]
Ok, here's what I'm going to do.  For the moment, I'll declare that `evaluate()` is not supported on models with multiple outputs.  That's a feature that needs design, and I'd don't think we've done that design yet.  How do we match up labels to outputs?  How do we even know the same metric is appropriate for all outputs?  (It seems like someone started to implement proper support for this, but never finished.  `GeneratorEvaluator` takes arguments for `labels` and `outputs`, but then the outputs argument is ignored and never used.)

I'll make evaluation work correctly for the single output case.  And I'll revise both tests so they don't use Databag and so the models they create have only one label and one output.

**Comment by peastman** [comment]
Tests are passing now.
